File: nodes/effects/delay.py
"""Delay effect node."""
import numpy as np
from core.node import Node
from core.port import AudioPort, ControlPort
from core import NodeRegistry


@NodeRegistry.register
class Delay(Node):
    category = "Effects"
    description = "Digital delay with feedback"

    def _setup_ports(self):
        self._inputs["audio"] = AudioPort("audio", self.id)
        self._outputs["audio"] = AudioPort("audio", self.id)

    def _setup_params(self):
        self._params["time"] = 0.3
        self._params["feedback"] = 0.4
        self._params["mix"] = 0.5
        self._params["_delay_buffer"] = np.zeros(44100 * 2, dtype=np.float32)
        self._params["_write_pos"] = 0

    def process(self, block_size: int) -> np.ndarray:
        audio_in = self._inputs["audio"].value
        if not hasattr(audio_in, '__len__') or len(audio_in) == 0:
            audio_in = np.zeros(block_size, dtype=np.float32)

        time_samples = int(self._params.get("time", 0.3) * 44100)
        feedback = self._params.get("feedback", 0.4)
        mix = self._params.get("mix", 0.5)
        delay_buffer = self._params.get("_delay_buffer", np.zeros(44100 * 2, dtype=np.float32))
        write_pos = self._params.get("_write_pos", 0)

        output = np.zeros(block_size, dtype=np.float32)
        for i in range(block_size):
            sample = audio_in[i] if i < len(audio_in) else 0.0
            read_pos = (write_pos - time_samples) % len(delay_buffer)
            delayed = delay_buffer[int(read_pos) % len(delay_buffer)]
            output[i] = sample * (1.0 - mix) + delayed * mix
            delay_buffer[int(write_pos) % len(delay_buffer)] = (
                sample + delayed * feedback
            )
            write_pos = (write_pos + 1) % len(delay_buffer)

        self._params["_delay_buffer"] = delay_buffer
        self._params["_write_pos"] = write_pos
        self._outputs["audio"].set_signal(output)
        return output

File: src/nodes/effects/delay.rs
use std::collections::HashMap;
use crate::core::{Node, NodeInfo, NodeParams, AudioBuffer, BlockSize, ParamValue, PortInfo, PortType};

pub struct DelayEffect {
    id: String,
    time: f64,
    feedback: f64,
    mix: f64,
    buffer: Vec<f32>,
    write_pos: usize,
}

impl DelayEffect {
    pub fn new(id: String) -> Self {
        Self { id, time: 0.3, feedback: 0.4, mix: 0.3, buffer: vec![0.0; 44100 * 2], write_pos: 0 }
    }
}

impl Node for DelayEffect {
    fn id(&self) -> &str { &self.id }
    fn name(&self) -> &str { "DelayEffect" }
    fn category(&self) -> &str { "Effect" }

    fn inputs(&self) -> Vec<PortInfo> {
        vec![PortInfo { name: "audio".to_string(), port_type: PortType::Audio }]
    }

    fn outputs(&self) -> Vec<PortInfo> {
        vec![PortInfo { name: "audio".to_string(), port_type: PortType::Audio }]
    }

    fn default_params(&self) -> NodeParams {
        let mut params = HashMap::new();
        params.insert("time".to_string(), ParamValue::Float(self.time));
        params.insert("feedback".to_string(), ParamValue::Float(self.feedback));
        params.insert("mix".to_string(), ParamValue::Float(self.mix));
        params
    }

    fn set_param(&mut self, name: &str, value: ParamValue) {
        let f = match value {
            ParamValue::Float(v) => v,
            ParamValue::Int(v) => v as f64,
            _ => return,
        };
        match name {
            "time" => { self.time = f.max(0.0).min(2.0); }
            "feedback" => { self.feedback = f.max(0.0).min(0.95); }
            "mix" => { self.mix = f.max(0.0).min(1.0); }
            _ => {}
        }
    }

    fn get_param(&self, name: &str) -> Option<ParamValue> {
        match name {
            "time" => Some(ParamValue::Float(self.time)),
            "feedback" => Some(ParamValue::Float(self.feedback)),
            "mix" => Some(ParamValue::Float(self.mix)),
            _ => None,
        }
    }

    fn process(&mut self, block_size: BlockSize, input: Option<&[f32]>) -> AudioBuffer {
        let delay_samples = (self.time * 44100.0) as usize;
        let feedback = self.feedback as f32;
        let mix = self.mix as f32;
        let mut buffer = AudioBuffer::with_capacity(block_size);

        for i in 0..block_size {
            let inp = input.and_then(|b| b.get(i)).copied().unwrap_or(0.0);
            let read_pos = (self.write_pos + self.buffer.len() - delay_samples) % self.buffer.len();
            let delayed = self.buffer[read_pos];
            self.buffer[self.write_pos] = inp + delayed * feedback;
            self.write_pos = (self.write_pos + 1) % self.buffer.len();
            let output = inp * (1.0 - mix) + delayed * mix;
            buffer.push(output);
        }

        buffer
    }

    fn to_info(&self, position: [f64; 2]) -> NodeInfo {
        NodeInfo {
            id: self.id.clone(), name: self.name().to_string(), category: self.category().to_string(),
            inputs: self.inputs(), outputs: self.outputs(), params: self.default_params(), position,
        node_kind: crate::core::NodeKind::Builtin,
        definition: None,
        }
    }
}
